guard-examples/cloudformation-resources/best-practices.guard

```diff
@@ -0,0 +1,67 @@
+# Auto-generated names are a CloudFormation best practice.
+# Resource Providers can use generateResourceIdentifier from IdentifierUtils (or equivalent)
+# This does not fail (yet?) for places where ${Resource}Name is a required property
+rule ensure_name_is_optional when required !empty {
+    required[*] {
+        this != 'Name'
+    }
+}
```

Comment:
For more context if curious, actually marked many as `required` as a mitigation because while most got the memo to generate identifiers to make them optional, some forgot to also return them in the `CreateHandler` `ResourceModel`, so updating/deleting would fail due to not having the correct identifier in the `ResourceModel`

Something to keep in mind for anyone generating identifiers to make them optional

Reply:
Shouldn't the contract test catch that (read after create being different)?

I could expand the rule to make sure that at least one of these is true

- Name is part of the primary identifier
- Name is part of an additional identifier
- Name is a createOnly property
- There is no updateHandler

(and if all current passing schemas still pass with these rules, I'd probably do that)

However, since you're not required to put Name in an identifier (or even make it require recreation), that might increase the number of false positives

Reply:
>Shouldn't the contract test catch that (read after create being different)?

Some types are still on previous frameworks where Registry contract tests aren't run, but even for Registry types where contract tests are run, it's difficult for Registry's framework to automatically distinguish optional properties where no value makes sense to return from optional properties where a value is generated in the handler or service

https://github.com/aws-cloudformation/cloudformation-resource-schema/issues/30#issuecomment-718288758